_posts/2018-06-14-fio-tutorial.md

```diff
@@ -0,0 +1,82 @@
+* numjobs parameter is telling how many threads/processes FIO will spawn to do the job; this is the best place to scale your workload,
+* blocksize parameter tells how big chunks of data will be transfered; this will influence bandwidth (bigger bs usually means higher BW/s) and IOPS (lower bs usually means higher IOPS).
+For details on available parameters and options go to [HOWTO page][191te3w1].
+
+FIO is generating I/O traffic using engines specific for the job. To specify which engine is used in a job there's workload's option `ioengine=my_engine`. All of them are described in [their c file][44ku0112] and all have corresponding [examples][55ku0123]. Few of them are related to persistent memory:
+** libpmem **
+This engine reads/writes data using libpmem library. Works on a namespace created in `fsdax` mode. Full exemplary workload for generating traffic of sequential reads using this engine can be found [here][fiolibPM]. There are additional comments within the jobfile to explain specific usage.
+
+** dev-dax **
+It also uses libpmem library, but as the name suggest it is specified to work with device-dax devices. Our full exemplary workload [DaxSeqR.fio][fiodevDX] shows how to properly use FIO with /dev/dax. Since we don't work on a "regular" file we use a little trick to achieve more realistic measurments - we set a seperate space for each thread doing reads/writes using option `offset_increment=(int)GB`. This way threads' requests do not overlap each other and results are not cached.
+
+** pmemblk **
+This engine is using libpmemblk library. Results delivered by this engine will not show you the best performance of your hardware, only what this specific library is capable of. Interestingly while using this engine, `blocksize` and `size` of a file are given as part of `filename` option, like here:
+{% highlight %}
+filename=/mnt/pmem6/testjob,512,1024000
+#size=1024000M
+#bs=512
+{% endhighlight %}
+Full exemplary workload doing traffic of sequential reads for pmemblk can be found [here][fioPMblk]
+
+** mmap **
+It's the most "basic" of mentioned engines, because its purpose is just to read from/write to a memory mapped region. It can be used with pmem, but is not tailor-made. It generates traffic doing memcpy to/from memory region. Again, full exemplary working workload is available: [MmapSeqRead.fio][fioMMap1]
+
+### Execution and results
+Command to run FIO is basically: `fio [options] [jobfile] ...`.
+Since we use workloads written to a file (in contrary to specifying parameters in command line), we execute FIO with command like `numactl -N 0 fio --output=my_workload.log --output-format=json my_workload.fio`. [numactl command][nctl1234] gurantees to map process to selected numa node. We can achieve this also by giving CPU mask to FIO option `cpus_allowed`. We also choose JSON format to save our results in it, because is much simpler to parse them afterwards.
```

Comment:
"in contract to" or "contrary to".
But here I would use "as opposed to"
"to pin processes to selected"
"We can achieve the same by assigning"

Why is it "much simpler"?